src/plugins/workspace/public/plugin.ts

```diff
@@ -11,59 +12,59 @@ import {
   AppMountParameters,
   AppNavLinkStatus,
 } from '../../../core/public';
-import { WORKSPACE_APP_ID, WORKSPACE_ID_IN_SESSION_STORAGE, PATHS } from '../common/constants';
+import { WORKSPACE_APP_ID, PATHS } from '../common/constants';
 import { WORKSPACE_ID_QUERYSTRING_NAME } from '../../../core/public';
 import { mountDropdownList } from './mount';
 
 export class WorkspacesPlugin implements Plugin<{}, {}> {
   private core?: CoreSetup;
-  private addWorkspaceListener() {
-    this.core?.workspaces.client.currentWorkspaceId$.subscribe((newWorkspaceId) => {
-      try {
-        sessionStorage.setItem(WORKSPACE_ID_IN_SESSION_STORAGE, newWorkspaceId);
-      } catch (e) {
-        /**
-         * in incognize mode, this method may throw an error
-         * */
-      }
-    });
-  }
   private getWorkpsaceIdFromQueryString(): string | null {
     const searchParams = new URLSearchParams(window.location.search);
     return searchParams.get(WORKSPACE_ID_QUERYSTRING_NAME);
   }
-  private getWorkpsaceIdFromSessionStorage(): string {
-    try {
-      return sessionStorage.getItem(WORKSPACE_ID_IN_SESSION_STORAGE) || '';
-    } catch (e) {
-      /**
-       * in incognize mode, this method may throw an error
-       * */
-      return '';
+  private async getWorkpsaceId(): Promise<string> {
+    if (this.getWorkpsaceIdFromQueryString()) {
+      return this.getWorkpsaceIdFromQueryString() || '';
     }
+
+    const currentWorkspaceIdResp = await this.core?.workspaces.client.getCurrentWorkspaceId();
+    if (currentWorkspaceIdResp?.success && currentWorkspaceIdResp?.result) {
+      return currentWorkspaceIdResp.result;
+    }
+
+    return '';
   }
-  private clearWorkspaceIdFromSessionStorage(): void {
-    try {
-      sessionStorage.removeItem(WORKSPACE_ID_IN_SESSION_STORAGE);
-    } catch (e) {
-      /**
-       * in incognize mode, this method may throw an error
-       * */
+  private async listenToApplicationChange(): Promise<void> {
```

Comment:
How do we deal with url change in the same application? Do we need to figure out this part?

Reply:
You mean `hash` change? As we are patch the workspaceId into the querystring, router inside plugins won't modify that part.

Reply:
Since we only subscribe application id change here. The URL won’t be updated if URL changed internally.

Reply:
I see, let me add some subscription on history change.

Reply:
I found an example in dashboard plugin. Here are the steps:
- Dashboard page has `_workspace_id_`
![image](https://github.com/ruanyl/OpenSearch-Dashboards/assets/4034161/3714bc1b-fb89-4352-9ab3-9f3d9b5269c2)
- Create dashboard page has `_workspace_id_`
![image](https://github.com/ruanyl/OpenSearch-Dashboards/assets/4034161/38482b94-063a-4e28-8b38-92befb965ad8)
- Create a Area visualize
![image](https://github.com/ruanyl/OpenSearch-Dashboards/assets/4034161/33217051-5b3d-4fb2-98e3-adbe7dbc3099)
- `_workspace_id_` missed in the visualize creating page
![image](https://github.com/ruanyl/OpenSearch-Dashboards/assets/4034161/925b4f2f-654c-4e8d-9bcb-40246c7ff6f6)

Reply:
Thanks, let me look into it.

Reply:
Add a hashchange listener to listen internal application change.